MyPy errors after MyPy upgrade

After MyPy was upgraded to fix an obscure issue for https://github.com/Tribler/py-ipv8/pull/1133, it found the following 7 new errors (and a lot of notes):

```
[test_ipv8_PR_mypy_check] $ /bin/bash /tmp/jenkins4071565667193143144.sh
ipv8/requestcache.py:99: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/requestcache.py:106: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/messaging/lazy_payload.py:68: error: Cannot assign to "__match_args__"  [misc]
ipv8/messaging/payload_dataclass.py:14: error: Accessing "__init__" on an instance is unsound, since instance.__init__ could be from an incompatible subclass  [misc]
ipv8/messaging/anonymization/payload.py:147: error: List item 0 has incompatible type "Type[IntroductionInfo]"; expected "Union[str, Serializable, List[FormatListType]]"  [list-item]
ipv8/community.py:458: error: Incompatible default for argument "from_peer" (default has type "None", argument has type "Peer")  [assignment]
ipv8/community.py:458: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
ipv8/community.py:458: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
ipv8/peerdiscovery/network.py:20: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/peerdiscovery/network.py:23: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/peerdiscovery/network.py:26: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/peerdiscovery/network.py:32: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/peerdiscovery/network.py:35: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/peerdiscovery/network.py:38: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/peerdiscovery/network.py:41: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
ipv8/attestation/identity/community.py:26: error: Incompatible default for argument "network" (default has type "None", argument has type "Network")  [assignment]
ipv8/attestation/identity/community.py:26: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
ipv8/attestation/identity/community.py:26: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
ipv8/attestation/identity/community.py:27: error: Incompatible default for argument "identity_manager" (default has type "None", argument has type "IdentityManager")  [assignment]
ipv8/attestation/identity/community.py:27: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
ipv8/attestation/identity/community.py:27: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
ipv8/attestation/communication_manager.py:194: error: Incompatible default for argument "working_directory" (default has type "None", argument has type "str")  [assignment]
ipv8/attestation/communication_manager.py:194: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
ipv8/attestation/communication_manager.py:194: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
Found 7 errors in 6 files (checked 226 source files)
```

At least the errors should be fixed. The notes should be looked at.

--- a/ipv8/attestation/communication_manager.py
+++ b/ipv8/attestation/communication_manager.py
@@ -191,7 +191,8 @@ class CommunicationChannel(object):
 
 class CommunicationManager(object):
 
-    def __init__(self, ipv8_instance: IPv8, pseudonym_folder: str = "pseudonyms", working_directory: str = None):
+    def __init__(self, ipv8_instance: IPv8, pseudonym_folder: str = "pseudonyms",
+                 working_directory: typing.Optional[str] = None):
         super(CommunicationManager, self).__init__()
 
         self.ipv8_instance = ipv8_instance

--- a/ipv8/attestation/identity/community.py
+++ b/ipv8/attestation/identity/community.py
@@ -23,8 +23,9 @@ class IdentityCommunity(Community):
 
     community_id = unhexlify('d5889074c1e4c50423cdb6e9307ee0ca5695ead7')
 
-    def __init__(self, my_peer: Peer, endpoint: Endpoint, network: Network = None, max_peers: int = DEFAULT_MAX_PEERS,
-                 anonymize: bool = True, identity_manager: IdentityManager = None, working_directory: str = "."):
+    def __init__(self, my_peer: Peer, endpoint: Endpoint, network: typing.Optional[Network] = None,
+                 max_peers: int = DEFAULT_MAX_PEERS, anonymize: bool = True,
+                 identity_manager: typing.Optional[IdentityManager] = None, working_directory: str = "."):
         if network is None:
             network = Network()
         super(IdentityCommunity, self).__init__(my_peer, endpoint, network, max_peers, anonymize)

--- a/ipv8/community.py
+++ b/ipv8/community.py
@@ -455,7 +455,7 @@ class Community(EZPackOverlay):
         packet = self.create_introduction_request(peer.address, new_style=peer.new_style_intro)
         self.endpoint.send(peer.address, packet)
 
-    def get_new_introduction(self, from_peer: Peer = None):
+    def get_new_introduction(self, from_peer: Optional[Peer] = None):
         """
         Get a new introduction, or bootstrap if there are no available peers.
         """

--- a/ipv8/messaging/lazy_payload.py
+++ b/ipv8/messaging/lazy_payload.py
@@ -65,7 +65,7 @@ class VariablePayload(Payload):
             raise KeyError("%s missing %d arguments!" % (self.__class__.__name__, len(args) - index))
         if kwargs:
             raise KeyError("%s has leftover keyword arguments: %s!" % (self.__class__.__name__, str(kwargs)))
-        self.__class__.__match_args__ = tuple(self.names)
+        setattr(self.__class__, "__match_args__", tuple(self.names))
 
     @classmethod
     def from_unpack_list(cls, *args) -> VariablePayload:

--- a/ipv8/messaging/payload_dataclass.py
+++ b/ipv8/messaging/payload_dataclass.py
@@ -11,7 +11,7 @@ def type_from_format(fmt: str) -> TypeVar:
     Convert a Serializer format directive to a type usable with @dataclass.
     """
     out = object.__new__(TypeVar)
-    out.__init__(fmt)
+    TypeVar.__init__(out, fmt)
     return out
 
 

--- a/ipv8/messaging/serialization.py
+++ b/ipv8/messaging/serialization.py
@@ -13,7 +13,7 @@ ADDRESS_TYPE_DOMAIN_NAME = 0x02
 ADDRESS_TYPE_IPV6 = 0x03
 
 
-FormatListType = typing.Union[str, "Serializable", typing.List["FormatListType"]]  # type:ignore
+FormatListType = typing.Union[str, typing.Type["Serializable"], typing.List["FormatListType"]]  # type:ignore
 
 
 class PackError(RuntimeError):

--- a/ipv8/peerdiscovery/network.py
+++ b/ipv8/peerdiscovery/network.py
@@ -16,7 +16,7 @@ WalkableAddress = namedtuple('WalkableAddress', ['introduced_by', 'services', 'n
 
 class Network(object):
 
-    def __init__(self):
+    def __init__(self) -> None:
         self._all_addresses: Dict[Address, WalkableAddress] = {}
         '''All known IP:port addresses, mapped to (introduction peer, services, new_style).'''
 
@@ -42,18 +42,18 @@ class Network(object):
         '''Map of service identifiers to local overlays.'''
 
         self.reverse_ip_cache_size = 500
-        self.reverse_ip_lookup = OrderedDict()
+        self.reverse_ip_lookup: OrderedDict[Address, Peer] = OrderedDict()
         '''Cache of IP:port -> Peer. This is a cache rather than a normal dictionary (the addresses of a peer are
         temporal and can grow infinitely): we rotate out old information to avoid a memory leak.'''
 
         self.reverse_intro_cache_size = 500
-        self.reverse_intro_lookup = OrderedDict()
+        self.reverse_intro_lookup: OrderedDict[Peer, List[Address]] = OrderedDict()
         '''Map of Peer -> [IP:port], reversing the information from _all_addresses. This is a cache rather than a
         normal dictionary (the addresses of a peer are temporal and can grow infinitely): we rotate out old
         information to avoid a memory leak.'''
 
         self.reverse_service_cache_size = 500
-        self.reverse_service_lookup = OrderedDict()
+        self.reverse_service_lookup: OrderedDict[Service, List[Peer]] = OrderedDict()
         '''Cache of service_id -> [Peer]. This is a cache rather than a normal dictionary (the services of a peer may
         be temporal and can grow infinitely): we rotate out old information to avoid a memory leak.'''
 
@@ -217,7 +217,7 @@ class Network(object):
                 out = new_out
             return out
 
-    def get_verified_by_address(self, address: Address) -> Peer:
+    def get_verified_by_address(self, address: Address) -> Optional[Peer]:
         """
         Get a verified Peer by its IP address.
 

--- a/ipv8/requestcache.py
+++ b/ipv8/requestcache.py
@@ -3,7 +3,7 @@ from asyncio import CancelledError, gather
 from contextlib import contextmanager, suppress
 from random import random
 from threading import Lock
-from typing import Generator, Iterable, Optional, Type
+from typing import Dict, Generator, Iterable, Optional, Type
 
 from .taskmanager import TaskManager
 
@@ -84,7 +84,7 @@ class RandomNumberCache(NumberCache):
 
 class RequestCache(TaskManager):
 
-    def __init__(self):
+    def __init__(self) -> None:
         """
         Creates a new RequestCache instance.
         """
@@ -92,7 +92,7 @@ class RequestCache(TaskManager):
 
         self._logger = logging.getLogger(self.__class__.__name__)
 
-        self._identifiers = dict()
+        self._identifiers: Dict[str, NumberCache] = dict()
         self.lock = Lock()
         self._shutdown = False
 
@@ -110,7 +110,7 @@ class RequestCache(TaskManager):
         This is used internally for ``passthrough()``, don't modify this directly!
         """
 
-    def add(self, cache):
+    def add(self, cache: NumberCache):
         """
         Add CACHE into this RequestCache instance.
 

